### PR TITLE
Added Comment and CommentLinks

### PR DESCRIPTION
**Pull request by datanel**

### CODE REVIEW

File: src/utils.rs [hunk 1]
```diff
@@ -34,3 +34,18 @@ where
...
+{
+    use serde::Deserialize;
+    let s = String::deserialize(deserializer)?;
+    if s.is_empty() {
+        return Ok(Default::default());
+    }
+
+    ::serde_json::from_value(::serde_json::value::Value::String(s))
+        .map_err(::serde::de::Error::custom)
+}
```
Comment (TeXitoi) [inline src/utils.rs]: Inspired by https://github.com/BurntSushi/rust-csv/blob/master/src/lib.rs#L347

```rust
pub fn de_with_empty_default<'de, T: Default, D>(de: D) -> Result<T, D::Error>
where
    D: ::serde::Deserializer<'de>,
    for<'d> T: serde::Deserialize<'d>,
{
    use serde::Deserialize;
    Option::<T>::deserialize(de).map(|opt| opt.unwrap_or_else(Default::default))
}
```

File: Cargo.toml [hunk 2]
```diff
@@ -14,7 +14,7 @@ log = "0.4"
 get-corresponding-derive = { path = "get-corresponding-derive" }
 tempdir = "0.3.5"
 chrono = "0.4.0"
+serde_json = "1"
```
Comment (TeXitoi) [inline Cargo.toml]: You can revert that with the new de_with_empty_default

File: src/lib.rs [hunk 3]
```diff
@@ -8,6 +8,7 @@ extern crate log;
 extern crate serde;
 #[macro_use]
 extern crate serde_derive;
+extern crate serde_json;
```
Comment (TeXitoi) [inline src/lib.rs]: You can revert that too

File: src/ntfs.rs [hunk 4]
```diff
@@ -253,6 +256,60 @@ fn manage_feed_infos(collections: &mut Collections, path: &path::Path) {
...
+    collection
+        .index_mut(idx)
+        .comment_links_mut()
+        .push(comment_link.comment_id);
+}
+
+fn manage_comments(collections: &mut Collections, path: &path::Path) {
+    if csv::Reader::from_path(path.join("comments.txt")).is_ok() {
+        info!("Reading comments.txt");
+        collections.comments = make_collection(path, "comments.txt");
```
Comment (TeXitoi) [inline src/ntfs.rs]: better to have an optional version of make_collection, that try to create a Read, and then, if successful, create a csv::Reader::from_reader. Creating a csv::Reader for cheecking the presence of the file is a bit overkill.

File: src/ntfs.rs [hunk 5]
```diff
@@ -253,6 +256,60 @@ fn manage_feed_infos(collections: &mut Collections, path: &path::Path) {
...
+                let comment_link: CommentLink = comment_link;
+                match comment_link.object_type.as_str() {
+                    "stop_area" => insert_comment_link(&mut collections.stop_areas, comment_link),
+                    "stop_point" => insert_comment_link(&mut collections.stop_points, comment_link),
+                    "line" => insert_comment_link(&mut collections.lines, comment_link),
+                    "route" => insert_comment_link(&mut collections.routes, comment_link),
+                    "trip" => insert_comment_link(&mut collections.vehicle_journeys, comment_link),
+                    "stop_time" => {
+                        insert_comment_link(&mut collections.vehicle_journeys, comment_link)
+                    }
```
Comment (TeXitoi) [inline src/ntfs.rs]: stop_time should be on a stop_time, not a vehicle journey. You'll have id mismatch like that.

File: src/objects.rs [hunk 6]
```diff
@@ -306,7 +331,9 @@ pub struct StopTime {
     pub dropoff_type: u8,
     pub datetime_estimated: bool,
     pub local_zone_id: Option<u16>,
+    pub comment_links: CommentLinksT,
```
Comment (TeXitoi) [inline src/objects.rs]: Still RAM problem with that.

File: src/objects.rs [hunk 7]
```diff
@@ -400,6 +431,33 @@ impl Id<Company> for Company {
...
+pub enum CommentType {
+    Information,
+    OnDemandTransport,
+}
+
+impl Default for CommentType {
+    fn default() -> Self {
+        CommentType::Information
+    }
+}
```
Comment (TeXitoi) [inline src/objects.rs]: you can use https://crates.io/crates/derivative

File: src/objects.rs (hunk 6 above)
Comment (datanel) [reply src/objects.rs]: absolutely I remove it?

File: src/ntfs.rs (hunk 5 above)
Comment (datanel) [reply src/ntfs.rs]: oups

File: src/objects.rs (hunk 6 above)
Comment (TeXitoi) [reply src/objects.rs]: Yeah, we'll handle all of that later

File: src/ntfs.rs (hunk 4 above)
Comment (TeXitoi) [reply src/ntfs.rs]: a make_collection_from_reader can be great

Comment (datanel) [reply src/ntfs.rs]: or I can do `if path.join("comments.txt").exists() {`

Comment (TeXitoi) [reply src/ntfs.rs]: looks good, yes.

Comment (TeXitoi) [reply src/ntfs.rs]: even better: `is_file`

Comment (TeXitoi) [reply src/ntfs.rs]: maybe not, as you wish.

File: src/ntfs.rs [hunk 8]
```diff
@@ -253,6 +255,61 @@ fn manage_feed_infos(collections: &mut Collections, path: &path::Path) {
...
+                    "stop_point" => insert_comment_link(&mut collections.stop_points, comment_link),
+                    "line" => insert_comment_link(&mut collections.lines, comment_link),
+                    "route" => insert_comment_link(&mut collections.routes, comment_link),
+                    "trip" => insert_comment_link(&mut collections.vehicle_journeys, comment_link),
+                    // TODO find something better
+                    // "stop_time" => {
+                    //     insert_comment_link(&mut collections.stop_times, comment_link)
+                    // }
+                    // TODO
+                    // "line_group" => insert_comment_link(&mut collections.line_groups, code),
```
Comment (TeXitoi) [inline src/ntfs.rs]: you may want to handle the case with a `warn!()` message to not crash if you have data with stop_time or line_group comment.

**Pull request merged**